Catalog3 3

Neenjahw opened the pull request:
2/3 Эпика на ревью
Архитектура - MVP
Способ верстки - кодом
ссылка на таск трекер - https://github.com/users/gimon2000/projects/3/views/1
Ссылка на декомпозицию - https://github.com/Neenjahw/iOS-FakeNFT-StarterProject-Public/blob/main/Decomposition/Catalog.md

Реальное время, затраченное на реализацию 3/3 эпика:
Реализовать сортировку Экрана каталога - 1 час
Реализовать добавлении NFT в избранное - 3 часа
Реализовать добавление NFT в корзину - 6 часов

Neenjahw committed:
1. Исправил расстояние между ячейками таблицы на экране каталога коллекций.
2. Реализовал сортировку коллекций NFT согласно макета.

Neenjahw committed:
Перенес логику сортировки коллекций в NftCatalogPresenter

Neenjahw committed:
Реализовал добавление NFT в избранное и корзину

Helvetii951 commented on FakeNFT/Scenes /Catalog/NftCatalogPresenter.swift:
```diff
@@ -4,13 +4,20 @@ import Foundation
+    case byName
+    case byNftCount
+}
+
```
Стоит вынести енум в отдельный файл. Может еще где-то пригодиться.

Helvetii951 commented on FakeNFT/Scenes /Catalog/NftCatalogViewController.swift:
```diff
@@ -43,16 +52,42 @@ final class NftCatalogViewController: UIViewController {
+        }
+        
+        let cancelAction = UIAlertAction(title: Constants.alertCancel, style: .cancel)
+        
```
`alertAction1` лучше поменять на более понятное название. Чуть больше времени потратишь, но потом тебе и коллегам будет понятнее что происходит.

Helvetii951 commented on FakeNFT/Services/Requests/EditOrderRequest.swift:
```diff
@@ -0,0 +1,38 @@
+struct EditOrderRequest: NetworkRequest {
+    var endpoint: URL? {
+        URL(string: "\(RequestConstants.baseURL)/api/v1/orders/1")
+    }
```
Пути до конечной точки `URL`  желательно хранить тоже в константах. Так будет легко быстро найти и исправить что нужно.

Helvetii951 commented on FakeNFT/Foundation/MemoryStorage/OrderStorage.swift:
```diff
@@ -0,0 +1,24 @@
+    func getOrder() -> Order?
+}
+
+final class OrderStorageImpl: OrderStorage {
+    private var order: Order?
+    
+    private let syncQueue = DispatchQueue(label: "sync-order-queue")
+    
+    func saveOrder(_ order: Order) {
+        syncQueue.async { [weak self] in
+            self?.order = order
+        }
+    }
```
В нашем случае выполнение этой операции займет не так много времени. НО. За старания лукас👍 Лучше перестраховаться и создать отдельную очередь.

Helvetii951 commented on FakeNFT/Resources/Localizable.strings:
```diff
@@ -4,3 +4,7 @@
 "Error.unknown" = "Произошла неизвестная ошибка";
 "Error.repeat" = "Повторить";
 "Error.title" = "Ошибка";
+"Catalog.sort" = "Сортировка";
+"Catalog.sortByName" = "По названию";
+"Catalog.sortByNftCount" = "По количеству NFT";
+"Catalog.cancel" = "Закрыть";
```
Рекомендую называть локализованные строки так:
```
"module_name.screen_name.exact_string_name" = "some string";
```

Neenjahw committed:
1. Изменил contentMode у картинки коллекции на главном экране.
2. Добавил сохранение порядка сортировки между запусками.
3. Добавил переход на экран автора коллекции.

Neenjahw committed:
1. Вынес SortOption в отдельный файл.
2. Переименовал title у action в alertViewController на главном экране.
3. Перенес пути до конечной точки URL в константы.
